### PR TITLE
Fix/hash in value

### PR DESCRIPTION
**Pull request by warjiang**
1. add .gitignore in case of ide of jetbrains
2. fix case of '#' as part of value, not as remark of comment
```
Host wap1 # there is a proxy available for this one...
    UserKnownHostsFile /dev/null
    StrictHostKeyChecking no
    Port 22
    User user_name
    IdentityFile path_to_private_key
    ProxyCommand ncat --proxy-type socks5 --proxy-auth 'user_name:pwd@#123_^&*' --proxy proxy_ip:proxy_port %h %p
```
becase '#' in the value of ProxyCommand, the part behine '#' will be treated as comment, which cased truncate of value of ProxyCommand

### CODE REVIEW

File: lexer.go [hunk 1]
```diff
@@ -97,9 +103,39 @@ func (s *sshLexer) lexRvalue() sshLexStateFn {
 			s.skip()
 			return s.lexVoid
 		case '#':
+			if !isSingleQuoted && !isDoubleQuoted {
```
Comment (lonnywong) [inline lexer.go]: I think this is no longer needed.

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+# _build/
+
+# Files generated by popular Go packages (remove the comment below to include it)
+# node_modules/
+# *.log
+
+# Don't ignore .gitignore and .gitattributes files
+!.gitignore
+!.gitattributes
+demo
```
Comment (lonnywong) [inline .gitignore]: Please add a newline at the end of the file, including the following test files.

File: lexer.go [hunk 3]
```diff
@@ -97,9 +103,39 @@ func (s *sshLexer) lexRvalue() sshLexStateFn {
...
+			break
+		}
+		growingString += string(next)
+		s.next()
+	}
+	s.emit(tokenEOF)
+	return nil
+}
+
+func (s *sshLexer) lexRReservedvalue() sshLexStateFn {
```
Comment (lonnywong) [inline lexer.go]: I think there is no need to add a new function, just add an argument, such as:
```go
func (s *sshLexer) lexRvalue(noComment bool) sshLexStateFn {
  for {
    switch next {
    case '#':
      if !noComment {
        // ...
      }
    }
  }
}
```

**Review by lonnywong — CHANGES_REQUESTED** [review]

File: lexer.go (hunk 1 above)
Comment (warjiang) [reply lexer.go]: In fact it keeps the same grammer as openssh. Is it ok to provide an alternative way for developer to choose 😆

Comment (lonnywong) [reply lexer.go]: I don't quite understand the logic of openssh. When is a comment and when isn’t it? I can't tell yet whether this modification is appropriate.

Comment (warjiang) [reply lexer.go]: +1, I cannot describe clear behaviour of openssh. It looks like a duck type. I decide to remove handle of single quote and double quote currently. I'll reopen another pr after  I have a deep learning of openssh.

File: lexer.go (hunk 3 above)
Comment (warjiang) [reply lexer.go]: I change `noComment `  to  `ignoreHash` for more human readable.

Comment (lonnywong) [reply lexer.go]: > I change `noComment ` to `ignoreHash` for more human readable.

Is `notEndWithComment` better?

File: lexer.go [hunk 4]
```diff
@@ -81,36 +81,43 @@ func (s *sshLexer) lexKey() sshLexStateFn {
...
 }
 
-func (s *sshLexer) lexRvalue() sshLexStateFn {
-	growingString := ""
-	for {
-		next := s.peek()
-		switch next {
-		case '\r':
-			if s.follow("\r\n") {
+func (s *sshLexer) lexRvalue(ignoreComment bool) sshLexStateFn {
```
Comment (lonnywong) [inline lexer.go]: Is `notEndWithComment` better?

Comment (warjiang) [reply lexer.go]: done

**Pull request merged**